src/main/java/net/neoforged/neoforge/tooltip/VanillaTooltipProviders.java

```diff
@@ -0,0 +1,263 @@
+                            flag1.setFalse();
+                        }
+
+                        addModifierTooltip(holder, adder, player, attributeHolder, attributeModifier);
+                    });
+                }
+            }
+        });
+
+        event.registerBelow(ITEM_ATTRIBUTES, ITEM_ADVENTURE_MODE_PREDICATES, (holder, player, context, adder, flag) -> {
+            var canBreak = holder.get(DataComponents.CAN_BREAK);
+
+            if (canBreak != null && canBreak.showInTooltip()) {
+                adder.accept(CommonComponents.EMPTY);
+                adder.accept(AdventureModePredicate.CAN_BREAK_HEADER);
+                canBreak.addToTooltip(adder);
+            }
+
+            var canPlaceOn = holder.get(DataComponents.CAN_PLACE_ON);
+
+            if (canPlaceOn != null && canPlaceOn.showInTooltip()) {
+                adder.accept(CommonComponents.EMPTY);
+                adder.accept(AdventureModePredicate.CAN_PLACE_HEADER);
+                canPlaceOn.addToTooltip(adder);
+            }
+        });
+
+        event.registerBelowAll(DISABLED, (holder, player, context, adder, flag) -> {
+            if (player != null && !holder.isItemEnabled(player.level().enabledFeatures()))
+                adder.accept(Component.translatable("item.disabled").withStyle(ChatFormatting.RED));
+        });
+
+        event.registerAbove(DISABLED, DATA_COMPONENT_COUNT, (holder, player, context, adder, flag) -> {
+            if (flag.isAdvanced()) {
+                var componentCount = holder.getComponents().size();
+
+                if (componentCount > 0)
+                    adder.accept(Component.translatable("item.components", componentCount).withStyle(ChatFormatting.DARK_GRAY));
+            }
+        });
+
+        event.registerAbove(DATA_COMPONENT_COUNT, ITEM_ID, (holder, player, context, adder, flag) -> {
+            if (flag.isAdvanced())
+                adder.accept(Component.literal(Util.getRegisteredName(BuiltInRegistries.ITEM, holder.getItem())).withStyle(ChatFormatting.DARK_GRAY));
+        });
+
+        event.registerAbove(ITEM_ID, ITEM_DURABILITY, (holder, player, context, adder, flag) -> {
+            if (flag.isAdvanced() && holder.isDamaged()) {
+                var maxDamage = holder.getMaxDamage();
+                adder.accept(Component.translatable("item.durability", maxDamage - holder.getDamageValue(), maxDamage));
+            }
+        });
+    }
+
+    private static ResourceLocation key(String identifier) {
+        return new ResourceLocation(ResourceLocation.DEFAULT_NAMESPACE, identifier);
+    }
+
+    private static void addModifierTooltip(ItemStack stack, Consumer<Component> adder, @Nullable Player player, Holder<Attribute> attribute, AttributeModifier attributeModifier) {
+        // Copied from ItemStack | private in vanilla
+        var d0 = attributeModifier.amount();
```

Comment:
I don't mind copy paste, but it should generally be updated with actual meaningful names